xtuner/utils/fileio.py

```diff
@@ -0,0 +1,179 @@
+        if isinstance(backend, LocalBackend):
+            return mkdir._fallback(path, *args, **kwargs)
+
+    @patch_func(os, 'makedirs')
+    def makedirs(path, *args, **kwargs):
+        backend = get_file_backend(path)
+        if isinstance(backend, LocalBackend):
+            return makedirs._fallback(path, *args, **kwargs)
+
+    @patch_func(os, 'listdir')
+    def listdir(path):
+        backend = get_file_backend(path)
+        if isinstance(backend, LocalBackend):
+            return listdir._fallback(path)
+        return backend.list_dir_or_file(path)
+
+    import filecmp
+
+    @patch_func(filecmp, 'cmp')
+    def cmp(f1, f2, *args, **kwargs):
+        with fileio.get_local_path(f1) as f1, fileio.get_local_path(f2) as f2:
+            return cmp._fallback(f1, f2, *args, **kwargs)
+
+    import shutil
+
+    @patch_func(shutil, 'copy')
+    def copy(src, dst, **kwargs):
+        backend = get_file_backend(src)
+        if isinstance(backend, LocalBackend):
+            return copy._fallback(src, dst, **kwargs)
+        return backend.copyfile_to_local(str(src), str(dst))
+
+    import torch
+
+    @patch_func(torch, 'load')
+    def load(f, *args, **kwargs):
+        if isinstance(f, str):
+            f = io.BytesIO(fileio.get(f))
+        return load._fallback(f, *args, **kwargs)
+
+    @patch_func(torch, 'save')
+    def save(obj, f, *args, **kwargs):
+        backend = get_file_backend(f)
+        if isinstance(backend, LocalBackend):
+            return save._fallback(obj, f, *args, **kwargs)
+
+        assert isinstance(f, str)
+        with io.BytesIO() as _f:
+            save._fallback(obj, _f, *args, **kwargs)
+            backend.put(_f.getvalue(), f)
+
+    from sentencepiece import SentencePieceProcessor
+
+    @patch_func(SentencePieceProcessor, 'LoadFromFile')
+    def LoadFromFile(cls, path):
+        if path:
+            backend = get_file_backend(path)
+            if isinstance(backend, LocalBackend):
+                return LoadFromFile._fallback(cls, path)
+            from tempfile import TemporaryDirectory
+            with TemporaryDirectory() as tmpdir:
+                local_path = backend.copyfile_to_local(path, tmpdir)
+                loaded_file = LoadFromFile._fallback(cls, local_path)
+            return loaded_file
+        else:
+            return LoadFromFile._fallback(cls, path)
+
+    try:
+        setattr(patch_fileio, '_patched', True)
+        yield
+    finally:
+        for patched_fn in patch_func._backup:
+            (module, fn_name_to_wrap, fn_to_wrap) = patched_fn
+            setattr(module, fn_name_to_wrap, fn_to_wrap)
+        if global_vars is not None and 'open' in global_vars:
+            global_vars['open'] = bak_open
+        setattr(patch_fileio, '_patched', False)
+
+
+def patch_hf_auto_from_pretrained():
+    if hasattr(patch_hf_auto_from_pretrained, '_patched'):
+        return
+
+    from transformers import AutoModelForCausalLM, AutoTokenizer
```

Comment:
Suggest modifying the `from_pretrained` method of these classes.

```python
from transformers import AutoConfig, AutoTokenizer, AutoImageProcessor, AutoFeatureExtractor, AutoProcessor, PreTrainedTokenizerBase, PreTrainedModel, ImageProcessingMixin, ProcessorMixin
from transformers.models.auto.auto_factory import _BaseAutoModelClass
from peft import PeftModel
```

This is an inflexible approach, but I can't think of a better solution.

Reply:
Done